enhanced/scraper.go

```diff
@@ -59,60 +59,72 @@ func (s *scraper) start(ctx context.Context, interval time.Duration, ch chan<- m
 
 // scrape performs a single scrape.
 func (s *scraper) scrape(ctx context.Context) map[string][]prometheus.Metric {
-	input := &cloudwatchlogs.FilterLogEventsInput{
-		LogGroupName:   aws.String("RDSOSMetrics"),
-		LogStreamNames: aws.StringSlice(s.logStreamNames),
-		StartTime:      aws.Int64(aws.TimeUnixMilli(s.nextStartTime)),
-	}
-	s.logger.Debugf("Requesting metrics since %s (last %s).", s.nextStartTime.UTC(), time.Since(s.nextStartTime))
 
-	// collect all returned events and metrics
 	allMetrics := make(map[string]map[time.Time][]prometheus.Metric) // ResourceID -> event timestamp -> metrics
-	collectAllMetrics := func(output *cloudwatchlogs.FilterLogEventsOutput, lastPage bool) bool {
-		for _, event := range output.Events {
-			l := s.logger.With("EventId", *event.EventId).With("LogStreamName", *event.LogStreamName)
-			l = l.With("Timestamp", aws.MillisecondsTimeValue(event.Timestamp).UTC())
-			l = l.With("IngestionTime", aws.MillisecondsTimeValue(event.IngestionTime).UTC())
-
-			var instance *sessions.Instance
-			for _, i := range s.instances {
-				if i.ResourceID == *event.LogStreamName {
-					instance = &i
-					break
+
+	// LogStreamNames parameter supports up to 100 items.
+	// https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_FilterLogEvents.html
+	for i := 0; len(s.logStreamNames)-(100*i) > 0; i++ {
```

Comment:
you might want to move this `len` call to a separate variable just to be safe, the compiler might not be able to optimize this, and if it cannot you'd be counting the stream on every iteration of the loop. I guess a simpler version of it would be something along the lines of this:

```suggestion
	streamCount := len(s.logStreamNames)
	batchSize := 100
	batchCount := (streamCount / batchSize) + 1
	for i := 0; i < batchCount; i++ {
```

Reply:
Better to just do `i := 0; i < len(s.logStreamNames; i += 100 { … }` and let the computer do the maths for you. 😉

Reply:
Oh, as an aside, unlike some languages, in Go the `len()` function does not actually count the elements of a slice in `O(n)` time. It just reads the `Len` field of the slice header: https://golang.org/pkg/reflect/#SliceHeader which is `O(1)` time.